[r2r] lightning ordermatching wip + library updates and more unit tests

shamardy opened the pull request:
Intermediate PR as part of #1045 
- Updated rust-lightning to v0.0.113
- Added channel `current_confirmations` and `required_confirmations` (which are easy to retrieve after the above update) to channel details RPCs. 
- Used `uuid` for internal channel id instead of `u64` after rust-lightning update since they now use `u128` for `user_channel_id`
- Updated `uuid` lib to latest version to use conversion from/to `u128` features
- Added unit tests for multi path payments.
- Added unit tests for claiming swaps on-chain for closed channels which is not currently working. After some discussion with rust-lightning devs this will be implemented in rust-lightning instead of AtomicDEX https://github.com/lightningdevkit/rust-lightning/issues/2017
- Started working in lightning ordermatching, `protocol_info` fields are used to check if an order can be matched or not depending on if a route is found between maker/taker for a lightning payment.
- Fixed 2 issues I discovered while executing a KMD/LNBTC swap, they were:
   - When electrums were down, if a channel was closed, the channel closing transaction wasn't broadcasted. Now I check for a network error and rebroadcast transactions after a delay if there was a network error.
   - If an invoice payment failed, if I retry paying the same invoice later, the payment wasn't updated to successful in DB if it were. Now I added a method to update the payment in the DB to update it in this case.

To be done in next PRs:
- Complete lightning order matching implementation and balance checks.
   - Right now inbound liquidity is not checked when placing an order to receive lightning BTC.
   - Show lightning node address in orderbook/best_orders response.
   - Fees can't be calculated accurately before an order is placed since they depend on the route. Will need to find a workaround for this.
- After updating to v0.0.113, an async event handler can be used instead of background processor, this paves the way for lightning to work in WASM.

ozkanonur reviewed (changes requested):
Amazing work! My notes from the partial review

shamardy commented on Cargo.lock:
```diff
@@ -456,6 +456,12 @@ version = "0.8.1"
+name = "bech32"
+version = "0.9.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "d86b93f97252c47b41663388e6d155714a9d0c398b99f1005cbc5f978b29f445"
```
Thanks a lot for upgrading `librustzcash`. Patch applied.

shamardy commented on mm2src/coins/lightning/ln_p2p.rs:
```diff
@@ -196,18 +197,34 @@ pub async fn init_peer_manager(
+        .expect(expect_msg)
+        .as_secs()
+        .try_into()
+        .expect(expect_msg);
```
Done

shamardy commented on mm2src/coins/lp_coins.rs:
```diff
@@ -2045,11 +2045,18 @@ pub trait MmCoin:
-    /// Get some of the coin config info in serialized format for p2p messaging.
-    fn coin_protocol_info(&self) -> Vec<u8>;
+    /// Get some of the coin protocol related info in serialized format for p2p messaging.
+    fn coin_protocol_info(&self, amount_to_receive: Option<MmNumber>) -> Vec<u8>;
```
`coin_protocol_info` first implementation used to return `Option<Vec<u8>>` and was changed in the following PR https://github.com/KomodoPlatform/atomicDEX-API/pull/1064 by Artem. We had some backward compatibility issues related to this before https://github.com/KomodoPlatform/atomicDEX-API/pull/1017, https://github.com/KomodoPlatform/atomicDEX-API/pull/1046 that would make me hesitant before changing this.

borngraced replied:
@shamardy I think in most cases, an empty `vec` == `None`...what I'm saying it's unnecessary to use `Option<Vec<_>>`  but instead just `Vec<_>`.

 seems this is how optional `keys` field are mostly stored in structures `mm2` so no problem ):

caglaryucekaya reviewed (commented):
Great work! Some questions and comments

ozkanonur reviewed (approved):
lgtm! just one suggestion

shamardy commented on mm2src/coins/lightning.rs:
```diff
@@ -361,7 +362,7 @@ impl LightningCoin {
             self.list_channels().await.into_iter().map(From::from).collect();
 
-        total_open_channels.sort_by(|a, b| a.rpc_channel_id.cmp(&b.rpc_channel_id));
+        total_open_channels.sort_by(|a, b| a.uuid.cmp(&b.uuid));
```
Thanks for noticing this. Before this PR, `rpc_channel_id` was incremented when a new channel is opened, so this was like sorting by timestamp which is the desired behaviour when calling this RPC since it's similar to other get by filter functions. Now I think it's not needed with uuids randomly generated, but since there are no timestamp for when a channel is open, I can probably use [`short_channel_id`](https://github.com/lightningdevkit/rust-lightning/blob/558b2f2904cc555a52bd57edbe3be44d8cfce6be/lightning/src/ln/channelmanager.rs#L1022) instead since it represents the channel's funding tx place in the chain which is akin to a timestamp. Will fix it.

shamardy replied:
Fixed

shamardy commented on mm2src/coins/lightning.rs:
```diff
@@ -1323,11 +1342,94 @@ impl MmCoin for LightningCoin {
+            .collect();
+        let node_id = PublicKeyForRPC(self.channel_manager.get_our_node_id());
+        let protocol_info = LightningProtocolInfo { node_id, route_hints };
+        rmp_serde::to_vec(&protocol_info).expect("Serialization should not fail")
```
Since `LightningProtocolInfo` is serializable, this should never fail.

shamardy commented on mm2src/coins/lightning.rs:
```diff
@@ -1323,11 +1342,94 @@ impl MmCoin for LightningCoin {
+            Some(Err(e)) => {
+                error!("{}", e);
+                return Vec::new();
+            },
```
If the amount can't be converted to `u64` it will be catched in previous balance checks anyways. Returning an error from this function for this conversion error will be overdoing this.

shamardy commented on mm2src/coins/lightning.rs:
```diff
@@ -1323,11 +1342,94 @@ impl MmCoin for LightningCoin {
+                },
+            };
+            route_hints.push(hint);
+        }
```
Done

shamardy commented on mm2src/coins/lightning/ln_serialization.rs:
```diff
@@ -101,7 +102,7 @@ impl<'de> de::Deserialize<'de> for PublicKeyForRPC {
 #[derive(Clone, Serialize)]
 pub struct ChannelDetailsForRPC {
-    pub rpc_channel_id: u64,
+    pub uuid: Uuid,
```
Done

shamardy commented:
I fixed 2 issues I discovered while executing a KMD/LNBTC swap, they were:
- When electrums were down, if a channel was closed, the channel closing transaction wasn't broadcasted. Now I check for a network error and rebroadcast transactions after a delay if there was a network error.
- If an invoice payment failed, if I retry paying the same invoice later, the payment wasn't updated to successful in DB if it were. Now I added a method to update the payment in the DB to update it in this case.

@caglaryucekaya @ozkanonur can you please review the new commits?

ozkanonur reviewed (approved):
> @ozkanonur can you please review the new commits?

Reviewed https://github.com/KomodoPlatform/atomicDEX-API/pull/1655/commits/f7953a718489e472046827fb2adefb42a04b0f89 and https://github.com/KomodoPlatform/atomicDEX-API/pull/1655/commits/e683bea2b034359f269a67430762bc4e36ec191a

Both looks good to me

shamardy commented:
@caglaryucekaya can you please approve if there are no more issues with this PR?

caglaryucekaya commented on mm2src/coins/lightning/ln_platform.rs:
```diff
@@ -601,15 +601,40 @@ impl BroadcasterInterface for Platform {
+                        if !e.get_inner().is_network_error() {
+                            break;
+                        }
+                        Timer::sleep(TRY_LOOP_INTERVAL).await;
```
Should this loop repeat infinitely if the network error keeps coming?

shamardy replied:
Yes for now. When broadcasting transaction through p2p network is implemented, a loop to check if the transaction is on-chain or not can be added instead.

caglaryucekaya commented on mm2src/coins/lightning/ln_platform.rs:
```diff
@@ -601,15 +601,40 @@ impl BroadcasterInterface for Platform {
+                        error!("Converting transaction to bytes error:{}", e);
+                        break;
+                    },
+                };
```
I think you can take this out of the loop, there's no need to repeat it in case of network failure

shamardy replied:
done

caglaryucekaya reviewed (commented):
LGTM! Just two comments

caglaryucekaya reviewed (dismissed):
Great work :fire:

borngraced commented on mm2src/mm2_main/src/database/my_swaps.rs:
```diff
@@ -57,7 +57,7 @@ fn insert_saved_swap_sql(swap: SavedSwap) -> Option<(&'static str, Vec<String>)>
 pub enum SelectRecentSwapsUuidsErr {
     Sql(SqlError),
-    Parse(uuid::parser::ParseError),
+    Parse(uuid::Error),
```
`uuid::Error` is used in multiple places already so it's better to import once

```rust
use uuid::Error as UuidError
```

shamardy replied:
Done

borngraced commented on mm2src/mm2_main/src/database/my_swaps.rs:
```diff
@@ -68,8 +68,8 @@ impl From<SqlError> for SelectRecentSwapsUuidsErr {
 
-impl From<uuid::parser::ParseError> for SelectRecentSwapsUuidsErr {
-    fn from(err: uuid::parser::ParseError) -> Self { SelectRecentSwapsUuidsErr::Parse(err) }
+impl From<uuid::Error> for SelectRecentSwapsUuidsErr {
```
`UuidError`

shamardy replied:
Done

borngraced reviewed (changes requested):
Great work! just a note and suggestion.
